test/services/optout/OptOutPropositionSpec.scala

```diff
@@ -0,0 +1,541 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package services.optout
+
+import models.incomeSourceDetails.TaxYear
+import models.itsaStatus.ITSAStatus.{Annual, ITSAStatus, Mandated, NoStatus, Voluntary}
+import org.scalatest.prop.TableDrivenPropertyChecks._
+import testUtils.UnitSpec
+
+import scala.io.Source
+
+class OptOutPropositionSpec extends UnitSpec {
+
+  "Parse opt out scenarios from tsv file" ignore {
+      /*  The programme has specified all required Opt Out scenarios in a large spreadsheet.
+          This code can ingest the data and convert to parameters for table based test found below.
+          This generator is currently ignored but can be reactivated to regenerate the test data if needed.
+            - To reactivate generator, replace "ignore" with "in"
+
+        Selected first 9 columns of Opt Out Scenarios Spreadsheet (currently v15).
+          - to avoid cells with carriage returns in to make parsing easier.
+        Paste into a fresh spreadsheet to save as OptOutScenarios.tsv in project root directory.
+          - csv has issues with commas within cells, therefore use tsv to allow simple parsing
+        Running test generates formatted test data in console, based on the required scenarios.
+          - Used to update 'scenarios' in table based test below.
+     */
+
+    val tsvOptOut = Source.fromFile("OptOutScenarios.tsv")
+
+    val lines = tsvOptOut.getLines().drop(2)
+
+    def parseTaxYears(taxYear: String): Seq[String] = {
+      taxYear match {
+        case "Nothing displayed" => Seq()
+        case "N/A" => Seq()
+        case _ => taxYear.replaceAll("\"", "").split("( [aA]nd |, )").map(_.trim).toSeq
+      }
+    }
+
+    def formatTaxYears(taxYears: Seq[String]) = {
+      "Seq(" ++ taxYears.map(option => s"\"$option\"").mkString(", ") ++ ")"
+    }
+
+    def parseIsValid(valid: String): Boolean = {
```

Comment:
I think this could be more robust regarding the input. The following would be preferable.
```scala
    def parseIsValid(valid: String): Boolean =
      valid match {
      case "Allowed" => true
      case "N/A"     => false
      case _         => throw new RuntimeException("Unexpected entry in Valid column")
    }
```